pkg/agent/agent.go

```diff
@@ -695,9 +703,15 @@ func (i *Initializer) configureGatewayInterface(gatewayIface *interfacestore.Int
 		klog.Errorf("Failed to find host link for gateway %s: %v", i.hostGateway, err)
 		return err
 	}
-
-	i.nodeConfig.GatewayConfig = &config.GatewayConfig{Name: i.hostGateway, MAC: gwMAC, OFPort: uint32(gatewayIface.OFPort)}
-	gatewayIface.MAC = gwMAC
+	i.nodeConfig.GatewayConfig = &config.GatewayConfig{Name: i.hostGateway, OFPort: uint32(gatewayIface.OFPort)}
+	// Use the MAC configured in the network interface when the gatewayIface.MAC is not set. This may
+	// happen in upgrade case.
+	if gatewayIface.MAC == nil {
+		gatewayIface.MAC = gwMAC
+		// Ignore the error as it does not impact the following logic even if no value is set in OVSDB.
+		i.ovsBridgeClient.SetInterfaceMAC(gatewayIface.InterfaceName, gwMAC)
+	}
```

Comment:
Perhaps the code could be unified as this:
```
// Persist the MAC configured in the network interface when the gatewayIface.MAC is not set. This may
// happen in upgrade case.
// Note the "mac" field in Windows OVS internal Interface has no impact on the network adapter's actual MAC, set it to the same value just to keep consistency.
if gatewayIface.MAC != gwMAC {
    gatewayIface.MAC = gwMAC
    if err := i.ovsBridgeClient.SetInterfaceMAC(gatewayIface.InterfaceName, gwMAC); err != nil {
        klog.ErrorS(err, "Failed to persist interface MAC address", "interface", gatewayIface.InterfaceName, "mac", gwMAC)
    }
}
i.nodeConfig.GatewayConfig = &config.GatewayConfig{Name: i.hostGateway, MAC: gwMAC, OFPort: uint32(gatewayIface.OFPort)}
```

Reply:
I don't think we can use `gatewayIface.MAC != gwMAC `, when we use a static MAC configuring internal OVS Interface (antrea-gw0), the network interface created on the host firstly use a random MAC, and then is updated to the desired address eventually. This is my observation.